{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "TCGA Glioma Clinical Data.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true,
      "authorship_tag": "ABX9TyOabKfwJPvGmua75mSCVX4H",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/sbpatel2009/NGS-Panel-Analyzer/blob/master/TCGA_Glioma_Clinical_Data.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "UGq6R0KRe3I6"
      },
      "source": [
        "# **INTRODUCTION**\n",
        "This file uploads, views, examines, and cleans the combined low grade glioma and glioblastoma clinical dataset generated by the TCGA.\n",
        "\n",
        "The clinical data was downloaded from cBioPortal:  https://www.cbioportal.org/study/clinicalData?id=lgggbm_tcga_pub\n",
        "\n",
        "**Relevant publications**\n",
        "\n",
        "1.   [Comprehensive genomic characterization defines human glioblastoma genes and core pathwaysExit Disclaimer\n",
        "Nature. 2008;455(7216):1061-1068. doi:10.1038/nature07385](https://www.nature.com/articles/nature07385)\n",
        "2.   [The somatic genomic landscape of glioblastomaExit Disclaimer Cell. 2013;155(2):462-477. doi:10.1016/j.cell.2013.09.034 ](https://www.cell.com/cell/fulltext/S0092-8674(13)01208-7)\n",
        "3.   [Comprehensive, Integrative Genomic Analysis of Diffuse Lower-Grade GliomasExit Disclaimer\n",
        "N Engl J Med. 2015;372(26):2481-2498. doi:10.1056/NEJMoa1402121 ](https://www.nejm.org/doi/full/10.1056/NEJMoa1402121)\n",
        "4.  [Molecular Profiling Reveals Biologically Discrete Subsets and Pathways of Progression in Diffuse Glioma.  Cell. 2016 Jan 28;164(3):550-63. doi: 10.1016/j.cell.2015.12.028.](https://www.nejm.org/doi/full/10.1056/NEJMoa1402121)"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "kk4zLiYogPOq",
        "outputId": "8ba5cec7-2634-42c2-c0f2-d7c62196bb55"
      },
      "source": [
        "!pip install scikit-survival  #https://scikit-survival.readthedocs.io/en/stable/index.html"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: scikit-survival in /usr/local/lib/python3.7/dist-packages (0.16.0)\n",
            "Requirement already satisfied: scipy!=1.3.0,>=1.0 in /usr/local/lib/python3.7/dist-packages (from scikit-survival) (1.4.1)\n",
            "Requirement already satisfied: pandas>=0.25 in /usr/local/lib/python3.7/dist-packages (from scikit-survival) (1.1.5)\n",
            "Requirement already satisfied: numexpr in /usr/local/lib/python3.7/dist-packages (from scikit-survival) (2.7.3)\n",
            "Requirement already satisfied: scikit-learn<0.25,>=0.24.0 in /usr/local/lib/python3.7/dist-packages (from scikit-survival) (0.24.2)\n",
            "Requirement already satisfied: ecos in /usr/local/lib/python3.7/dist-packages (from scikit-survival) (2.0.7.post1)\n",
            "Requirement already satisfied: joblib in /usr/local/lib/python3.7/dist-packages (from scikit-survival) (1.1.0)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from scikit-survival) (1.19.5)\n",
            "Requirement already satisfied: osqp!=0.6.0,!=0.6.1 in /usr/local/lib/python3.7/dist-packages (from scikit-survival) (0.6.2.post0)\n",
            "Requirement already satisfied: qdldl in /usr/local/lib/python3.7/dist-packages (from osqp!=0.6.0,!=0.6.1->scikit-survival) (0.1.5.post0)\n",
            "Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.7/dist-packages (from pandas>=0.25->scikit-survival) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2017.2 in /usr/local/lib/python3.7/dist-packages (from pandas>=0.25->scikit-survival) (2018.9)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.7/dist-packages (from python-dateutil>=2.7.3->pandas>=0.25->scikit-survival) (1.15.0)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from scikit-learn<0.25,>=0.24.0->scikit-survival) (3.0.0)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HKKMjcLcezHe"
      },
      "source": [
        "#LOAD THE RELEVANT PACKAGES\n",
        "# Import supporting modules\n",
        "import numpy as np\n",
        "import glob, os\n",
        "import pandas as pd\n",
        "import seaborn as sns\n",
        "%matplotlib inline\n",
        "import matplotlib.pylab as plt\n",
        "import sklearn  #  Scikit-learn, one of the best and most documented machine learning libaries for Python\n",
        "import sksurv"
      ],
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nA19JfmJuWIl"
      },
      "source": [
        "file_name = \"lgggbm_tcga_pub_clinical_data.tsv\"\n",
        "directory = \"https://drive.google.com/drive/folders/1W50b4nOhoVZv6JQDdZjteyfVuEU3La65\"\n",
        "file_path = directory + \"/\" + file_name\n",
        "url = \"https://drive.google.com/file/d/1J6vWyVk4_f83qrgFsMEu-kLZe775y7GA/view?usp=sharing\"\n",
        "data = pd.read_csv(filepath_or_buffer = url, sep = \"\\t\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "qoyUdFxawWIW"
      },
      "source": [
        "from google.colab import files\n",
        "uploaded = files.upload()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LbT53rkmwz8g"
      },
      "source": [
        "uploaded.keys()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "71-6sCMvwhnF"
      },
      "source": [
        "import io\n",
        "data = pd.read_csv(io.BytesIO(uploaded[\"lgggbm_tcga_pub_clinical_data.tsv\"]))\n",
        "data"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "YZZi442auzEe"
      },
      "source": [
        "data"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "KEXWSBZrsWNa"
      },
      "source": [
        "print(\"shape =\", data.shape)\n",
        "print(\"size =\", data.size)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-nb8_NIRsbMI"
      },
      "source": [
        "data.info()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LdULPn5xsdCp"
      },
      "source": [
        "data.describe()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "X46II--9sgRM"
      },
      "source": [
        "print(data.isna().sum() / data.shape[0] * 100)  #Print the percent of missing values per column"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "egKxSs05sisd"
      },
      "source": [
        "#Columns indicate whether or not a test was performd\n",
        "print(data[\"HM27\"].unique())\n",
        "print(data[\"HM450\"].unique())\n",
        "print(data[\"Whole Exome Sequencing\"].unique())\n",
        "print(data[\"Whole Genome Sequencing\"].unique())\n",
        "print(data[\"RPPA\"].unique())\n",
        "print(data[\"SNP6\"].unique())\n",
        "print(data[\"U133a\"].unique())\n",
        "print(data[\"RNA-SEQ Data\"].unique(), \"\\n\\n\")\n",
        "\n",
        "#Show number of unique values in each column and then identify the columns with only 1 unique value\n",
        "print(data.nunique(), \"\\n\\n\")\n",
        "print(data[data.nunique()[(data.nunique() == 1)].index])"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ivqAaKtJsmEl"
      },
      "source": [
        "fig, axes = plt.subplots(3, 3, figsize=(20, 15))\n",
        "fig.suptitle(\"Glioma Dataset Summary\")\n",
        "\n",
        "sns.distplot(ax = axes[0, 0], a = data[\"Diagnosis Age\"], kde = False)\n",
        "axes[0, 0].set_title(\"Age distribution\")\n",
        "axes[0, 0].set_ylabel(\"Number of patients\")\n",
        "\n",
        "axes[0, 1].pie(data[\"Sex\"].value_counts(), labels = data[\"Sex\"].value_counts().index, colors = sns.color_palette('pastel')[0:2], autopct='%.0f%%')\n",
        "axes[0, 1].set_title(\"Sex distribution\")\n",
        "\n",
        "sns.distplot(ax = axes[0, 2], a = data[\"Karnofsky Performance Score\"], kde = False)\n",
        "axes[0, 2].set_title(\"Karnofsky Performance Score\")\n",
        "axes[0, 2].set_ylabel(\"Number of patients\")\n",
        "\n",
        "axes[1, 0].pie(data[\"Neoplasm Histologic Type Name\"].value_counts(), labels = data[\"Neoplasm Histologic Type Name\"].value_counts().index, colors = sns.color_palette('pastel')[0:4], autopct='%.0f%%')\n",
        "axes[1, 0].set_title(\"Neoplasm Histologic Type Name\")\n",
        "\n",
        "axes[1, 1].pie(data[\"Neoplasm Histologic Grade\"].value_counts(), labels = data[\"Neoplasm Histologic Grade\"].value_counts().index, colors = sns.color_palette('pastel')[0:3], autopct='%.0f%%')\n",
        "axes[1, 1].set_title(\"Neoplasm Histologic Grade\")\n",
        "\n",
        "axes[1, 2].pie(data[\"Original Subtype\"].value_counts(), labels = data[\"Original Subtype\"].value_counts().index, colors = sns.color_palette('pastel')[0:9], autopct='%.0f%%')\n",
        "axes[1, 2].set_title(\"Original Subtype\")\n",
        "\n",
        "axes[2, 0].pie(data[\"IDH status\"].value_counts(), labels = data[\"IDH status\"].value_counts().index, colors = sns.color_palette('pastel')[0:4], autopct='%.0f%%')\n",
        "axes[2, 0].set_title(\"IDH status\")\n",
        "\n",
        "axes[2, 1].pie(data[\"IDH-1P10Q Subtype\"].value_counts(), labels = data[\"IDH-1P10Q Subtype\"].value_counts().index, colors = sns.color_palette('pastel')[0:4], autopct='%.0f%%')\n",
        "axes[2, 1].set_title(\"IDH-1P10Q Subtype\")\n",
        "\n",
        "axes[2, 2].pie(data[\"MGMT promoter status\"].value_counts(), labels = data[\"MGMT promoter status\"].value_counts().index, colors = sns.color_palette('pastel')[0:2], autopct='%.0f%%')\n",
        "axes[2, 2].set_title(\"MGMT promoter status\")\n",
        "\n",
        "plt.show()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "2Ec4dFaksodY"
      },
      "source": [
        "data[\"Overall Survival Status\"].replace(to_replace = \"1:DECEASED\", value = True, inplace = True)\n",
        "data[\"Overall Survival Status\"].replace(to_replace = \"0:LIVING\", value = False, inplace = True)\n",
        "data.dropna(axis = 0, inplace = True, subset = [\"Overall Survival Status\", \"Overall Survival (Months)\"])\n",
        "y = list(zip(data[\"Overall Survival Status\"], data[\"Overall Survival (Months)\"]))"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "lutg9JhjsqqH"
      },
      "source": [
        "categorical_columns = [\"Sex\", \"Neoplasm Histologic Type Name\", \"Neoplasm Histologic Grade\", \"Original Subtype\", \"IDH status\", \"MGMT promoter status\"]\n",
        "categorical_imputer = SimpleImputer(strategy = \"most_frequent\", copy = False)\n",
        "categorical_imputed = pd.DataFrame(categorical_imputer.fit_transform(data[categorical_columns]))\n",
        "categorical_imputed.columns = data[categorical_columns].columns\n",
        "categorical_imputed"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "eN5mr3BHssg_"
      },
      "source": [
        "numerical_columns = [\"Diagnosis Age\", \"Karnofsky Performance Score\"]\n",
        "numerical_imputer = SimpleImputer(strategy = \"mean\", copy = False)\n",
        "numerical_imputed = pd.DataFrame(numerical_imputer.fit_transform(data[numerical_columns]))\n",
        "numerical_imputed.columns = data[numerical_columns].columns\n",
        "num_normalized = pd.DataFrame(Normalizer().fit_transform(numerical_imputed))\n",
        "num_normalized.columns = numerical_imputed.columns\n",
        "num_normalized"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "g9XsDtN0svmU"
      },
      "source": [
        "X = pd.concat([categorical_imputed, num_normalized], axis = 1)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "7S05H_CqswOl"
      },
      "source": [
        "X"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "U-LxeCYWsw8Y"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "3UeqJEKJsye8"
      },
      "source": [
        "ordinal = OrdinalEncoder(categories = [[\"G2\", \"G3\", \"G4\"]])\n",
        "ordinalEnc = pd.DataFrame(ordinal.fit_transform(X.iloc[:, 2:3]))\n",
        "ordinalEnc.columns = X.iloc[:, 2:3].columns\n",
        "ordinalEnc"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HqCCpcN0s1HT"
      },
      "source": [
        "OHE = OneHotEncoder(sparse = False)\n",
        "OHencoded = pd.DataFrame(OHE.fit_transform(X.loc[:, [\"Sex\", \"Neoplasm Histologic Type Name\", \"Original Subtype\", \"IDH status\", \"MGMT promoter status\"]]))\n",
        "OHencoded"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "kTsixEshszH1"
      },
      "source": [
        "pd.concat([ordinalEnc, OHencoded, num_normalized], axis = 1)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Ut203A2Ls4uj"
      },
      "source": [
        "from sklearn.model_selection import cross_val_score\n",
        "\n",
        "numerical_columns = [\"Diagnosis Age\", \"Karnofsky Performance Score\"]\n",
        "nominal_columns = [\"Sex\", \"Neoplasm Histologic Type Name\", \"Original Subtype\", \"IDH status\", \"MGMT promoter status\"]\n",
        "ordinal_columns = [\"Neoplasm Histologic Grade\"]\n",
        "grade = [\"G2\", \"G3\", \"G4\"]\n",
        "\n",
        "numerical_transformer = Pipeline(steps = [\n",
        "    (\"imputer1\", SimpleImputer(strategy = \"mean\")),\n",
        "    (\"norm\", Normalizer())\n",
        "])\n",
        "\n",
        "nominal_transformer = Pipeline(steps = [\n",
        "    (\"imputer2\", SimpleImputer(strategy = \"most_frequent\")),\n",
        "    (\"OneHot\", OneHotEncoder())\n",
        "])\n",
        "\n",
        "ordinal_transformer = Pipeline(steps = [\n",
        "    (\"imputer3\", SimpleImputer(strategy = \"most_frequent\")),\n",
        "    (\"Ordinal\", OrdinalEncoder(categories = [grade]))\n",
        "])\n",
        "\n",
        "preprocessor = ColumnTransformer(transformers = [\n",
        "    (\"numerical\", numerical_transformer, numerical_columns),\n",
        "    (\"nominal\", nominal_transformer, nominal_columns),\n",
        "    (\"ordinal\", ordinal_transformer, ordinal_columns)\n",
        "])\n",
        "\n",
        "scores = -1 * cross_val_score(my_pipeline, X, y, CV = 5, scoring = \"neg_mean_absolute_error\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "XRPRIUr5s6-I"
      },
      "source": [
        "time, survival_prob = kaplan_meier_estimator(data[\"Overall Survival Status\"], data[\"Overall Survival (Months)\"])\n",
        "plt.step(time, survival_prob, where=\"post\")\n",
        "plt.ylabel(\"est. probability of survival $\\hat{S}(t)$\")\n",
        "plt.xlabel(\"time $t$\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sSMPZOo4s82t"
      },
      "source": [
        "pip install scikit-survival"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}